### PR TITLE
Fix workspace crates tests not being run

### PR DESCRIPTION
**Pull request by Urgau**
Fixes https://github.com/rust-lang/docs.rs/issues/2391

### CODE REVIEW

**Comment by Urgau** [comment]
@Nemo157 This PR fixes the tests of the `metadata` and `font-awesome-as-a-crate`, it build them on CI and adds them to the CI cache.

*I saw that the CI jobs were run on my fork as well so as bonus I also restricted CI to only run on the `master` branch (when pushing)*

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -2,6 +2,8 @@ name: CI
 
 on:
   push:
+    branches:
+      - master
```
Comment (syphar) [inline .github/workflows/ci.yml:6]: I don't fully remember this, but perhaps can you help me: 
would this mean when I have a branch on my fork, the tests wouldn't run there any more? 
Because that's a feature I'm using regularly :)

**Review by syphar — COMMENTED** [review]
thank you for working on this fix! 
I have one question about the branch limitation. 

Apart from that I'm actually not sure about the caching feature in general any more (which I added, so that's on me). It definitely improves CI times, but also it lead to us silently not running a part of the tests. And relying on test binary naming feels volatile at the least. 

we _could_ try to just use `Swatinem/rust-cache` in all jobs, and not share any binaries between the jobs. What do you think?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Urgau) [reply .github/workflows/ci.yml:6]: Yes, this would mean that CI won't be run on push for any branches other than `master` (fork or not).

The reason I added it, is that it's a waste of resource, particularly when doing a PR since every time a push happens CI is run on **both** the fork and here.

I would suggest doing PRs against your own fork to retain the capability of running CI in your fork.

**Comment by Urgau** [comment]
> Apart from that I'm actually not sure about the caching feature in general any more (which I added, so that's on me). It definitely improves CI times, but also it lead to us silently not running a part of the tests. And relying on test binary naming feels volatile at the least

Yeah, relying on test binary names is not great. I tried using `cargo --message-format` but this felt overkill.

> we _could_ try to just use `Swatinem/rust-cache` in all jobs, and not share any binaries between the jobs. What do you think?

Yes, that would be great. I can try to do it, but won't have time to do it until a few days.
Maybe we can merge this PR in the mean time and follow-up with this bigger refactor.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (syphar) [reply .github/workflows/ci.yml:6]: I would prefer the current setup where we have CI running on branches. 

Let's leave it out of this PR and revisit the question later.

**Comment by syphar** [comment]
> > Apart from that I'm actually not sure about the caching feature in general any more (which I added, so that's on me). It definitely improves CI times, but also it lead to us silently not running a part of the tests. And relying on test binary naming feels volatile at the least
> 
> Yeah, relying on test binary names is not great. I tried using `cargo --message-format` but this felt overkill.
> 
> > we _could_ try to just use `Swatinem/rust-cache` in all jobs, and not share any binaries between the jobs. What do you think?
> 
> Yes, that would be great. I can try to do it, but won't have time to do it until a few days. Maybe we can merge this PR in the mean time and follow-up with this bigger refactor.

This was broken for a long time, so having it broken for a little longer is not a big issue IMO. 
So IMO we should directly do the refactor (or rather, simplification, refactor sounds so huge 😄  ).

**Comment by syphar** [comment]
last week I also started doing this (dropping the binary caching), though I didn't dig through the remaining test failures yet: 

- https://github.com/rust-lang/docs.rs/compare/master...syphar:docs.rs:fix-test 
- https://github.com/syphar/docs.rs/actions/workflows/ci.yml?query=branch%3Afix-test

**Comment by Urgau** [comment]
I've updated the PR to remove the custom CI binary caching. I've consolidated the "build" and "build_test" jobs into a single "test" job. I also contains some tests fixes, but I did not fixed the ignored test since none are worth it, they mainly serve as copy/paste examples.

Let me if would like anything else.

@rustbot label -S-waiting-on-author +S-waiting-on-review

**Comment by Nemo157** [comment]
`#[ignored]` is really used as a way to categorize long-duration build-tests that we don't want to run locally normally, but they should be run in CI.

**Comment by Urgau** [comment]
> `#[ignored]` is really used as a way to categorize long-duration build-tests that we don't want to run locally normally, but they should be run in CI.

Oops, forgot this. Fixed.

**Comment by Urgau** [comment]
@Nemo157 @syphar What do you think about the latest changes? Can we move forward?

**Review by syphar — APPROVED** [review]
Sorry for the delay, vacation came in between. 

Thank you for the contribution!
